src/Analyzer/Passes/OptimizeRedundantFunctionsInOrderByPass.cpp

```diff
@@ -0,0 +1,111 @@
+#include <string_view>
+#include <Analyzer/Passes/OptimizeRedundantFunctionsInOrderByPass.h>
+#include <Analyzer/ColumnNode.h>
+#include <Analyzer/FunctionNode.h>
+#include <Analyzer/InDepthQueryTreeVisitor.h>
+#include <Analyzer/QueryNode.h>
+#include <Analyzer/SortNode.h>
+#include <Functions/IFunction.h>
+
+namespace DB
+{
+
+namespace
+{
+
+class OptimizeRedundantFunctionsInOrderByVisitor : public InDepthQueryTreeVisitor<OptimizeRedundantFunctionsInOrderByVisitor>
+{
+    std::unordered_set<std::string_view> existing_keys;
+
+    bool isRedundantExpression(FunctionNode * function)
+    {
+        if (function->getArguments().getNodes().empty())
+            return false;
+        const auto & function_base = function->getFunction();
+        if (!function_base || !function_base->isDeterministicInScopeOfQuery())
+            return false;
+
+        // TODO: handle constants here
```

Comment:
After recent refactoring we can see only `constant`, `column`, `function` nodes here.

Reply:
Yes. What do you want to say?
This comment is about optimizing `ORDER BY a, a + 1` to `ORDER BY a`. Currently, we don't support it, but in the future, it'd be nice to have it.